torchrl/envs/common.py

```diff
@@ -2874,12 +2888,20 @@ def __init__(
         self,
         *args,
         dtype: Optional[np.dtype] = None,
-        device: DEVICE_TYPING = None,
+        device: DEVICE_TYPING = NO_DEFAULT,
         batch_size: Optional[torch.Size] = None,
         allow_done_after_reset: bool = False,
         **kwargs,
     ):
-        if device is None:
+        if device is NO_DEFAULT:
+            warnings.warn(
+                "Your wrapper was not given a device. Currently, this "
+                "value will default to 'cpu'. From v0.5 it will "
+                "default to `None`. With a device of None, no device casting "
+                "is performed and the resulting tensordicts are deviceless. "
+                "Please set your device accordingly.",
+                category=DeprecationWarning,
```

Comment:
Asking for feedback on this

@matteobettini @skandermoalla @albertbou92 @BY571 @alexander-soare @dennismalmgren 

I know this is going to be massively annoying but we need to deprecate the default device of envs to CPU.
First it's arbitrary (an env can be on cuda by default) and restrictive (we now support envs that have parts of their input/output on cpu and part on another device).

The only option I see to avoid having this warning popping up all the time is to make an environment variable that you can set to use the non-deprecated behaviour. I personally don't like relying on environment variables...

Wdyt?

Reply:
I'm not sure I understand the problem.

Here's my understanding:
We want to deprecate the default device of an env.
If users don't want to do casting, with the current version they can already just not set the device (or set it to None?). This is valid behavior.
However, this will raise a (useless) warning until v0.5 and we want to avoid that.
Nevertheless, we want the warning to be raised when the user is not aware of the deprecation and just didn't put a device.

This is a new feature, so anyway users will be changing their code when adopting it, so an environment variable is not necessary I'd say. They can set an optional keyword argument `no_default_device=True` until v0.5.

"Please set your device accordingly." was a bit confusing. On my first read I thought that meant I have to set a device explicitly and None is not a device. On my second read, I think it just means I have to be careful when picking between a device and None.

Reply:
> However, this will raise a (useless) warning until v0.5 and we want to avoid that.
> Nevertheless, we want the warning to be raised when the user is not aware of the deprecation and just didn't put a device.

Yes so the only thing I want is that people who used to do `assert env.device == "cpu"` don't see their code breaking after 0.4, but they know that it will break in v0.5.

So I take it that we should keep the warning then, right? I should maybe use torchrl logger to allow people to disable warnings all at once.

Reply:
> Yes so the only thing I want is that people who used to do assert env.device == "cpu" don't see their code breaking after 0.4, but they know that it will break in v0.5.

This is what the current code does. So yes keep the warning.

For the early adopters, they will change their code anyway so it's fine to add an argument to suppress the warning IMO.